priUTR pipeline
Detect alternative 3' UTR in plants using RNA-seq data 

priUTR pipeline is developed in Scientific Linux release 6.10 (Carbon). This pipeline is based on Linux shell scripts, therefore it requires a 64-bit CPU computer running on Linux system. 5-50GB running RAM is recommended depending on data size. To make full use of standard RNA-seq data and to provide user-friendly 3’UTR analysis function for many non-model plant species and plant biologists who may have limited bioinformatic skills,  the priUTR pipeline can detect alternative 3’UTRs in comparison to the reference genome annotation and provide information regarding the identified alternative 3’UTRs. 

The program can be obtained from this Github website (https://github.com/mint1234/3UTR-).
Example input and output files can be downloaded from Waksman Institute of Microbiology, Rugters, The State University of New jersey (https://data.waksman.rutgers.edu/200106-3utr/).
For a detailed protocol to use the priUTR pipeline, please read the user manual at our Github website. 

Issues and bug reports
Please use https://github.com/mint1234/3UTR-/issues to submit issues, bug reports, and comments.

For further information, please contact with mintu@waksman.rutgers.edu.

