# priUTR pipeline
Detect alternative 3' UTR in plants using RNA-seq data 

priUTR pipeline is developed in Scientific Linux release 6.10 (Carbon). This pipeline is based on Linux shell scripts, therefore it requires a 64-bit CPU computer running on Linux system. 5-50GB running RAM is recommended depending on data size. And this pipeline was tested with the following version of the programs: 
-- samtools Version 1.3.1, 
http://www.htslib.org
-- bedtools v2.24.0, 
https://bedtools.readthedocs.io/en/latest/
-- TransDecoder Version 2.0.1, 
https://github.com/TransDecoder/TransDecoder/releases
-- gffread Version 0.9.8c, 
https://github.com/gpertea/gffread
-- ViennaRNA Version 2.4.14, 
https://www.tbi.univie.ac.at/RNA/RNAfold.1.html 
-- ToGenePred: 
https://bioconda-recipes-demo.readthedocs.io/en/docs/recipes/ucsc-gtftogenepred/README.html 
-- exteractFromFasta.pl: 
https://github.com/jonbra/NGS-Abel/blob/master/scripts/extractFromFasta.pl
-- qsub Version 3.0.4:
http://docs.adaptivecomputing.com/torque/2-5-12/help.htm#topics/commands/qsub.htm

User manual:
1.	Create three folders, namely “scripts”, “inputs” and “softwares”. Install all of the above-mentioned softwaresin the “softwares” folder. "scripts" folder contains three shell files: qsub-3r.sh, 3r.sh and genome.sh.
2.	Then move the following files into the “inputs” folder: (1) genome.gff3 (genome annotation); (2) genome.fa (assembled genome sequence); (3) the output files from RNA-seq read aligner and transcriptome assembler (Tophat and Cufflinks, etc.), including three “.bam” files corresponding to the triplicated RNA-seq samples respectively generated by Tophat, isoforms_fpkm.tracking (Cuffdiff output file to show isoform expression); merged.gtf (Cuffmerge output file of a transcript assembly merged from multiple RNA-seq samples).
3.	Please first set local path to “$PRE” in the shell file, and then execute the “genome.sh” file using command “sh genome.sh”; The script “genome.sh” is to extract 3’UTR information of all annotated transcripts in the reference genome. 
4.	Please set local path to “$PRE” in the ”qsub-3r.sh“ and ”3r.sh“ files. Then set condition names to “${SN}”, and set how many RNA-seq samples (or experiment treatments or conditions) to “i” in the ”qsub-3r.sh“ file. The priUTR pipeline can automatically loop execute for “i” times to proceed multiple RNA-seq samples (“i” samples, in this case) in one run. In the “3r.sh” file, the default setting of RPKM filter is “A=1”, and the default setting of 3’UTR-reads filter is “B=30”, they can be set or adjusted in the pipeline manually. Then execute the “qsub-3r.sh” using command “sh qsub-3r.sh” (which loop executes “3r.sh”). 

Results and outputs:
1.	The output results will be generated in two folders: “results” and “2nd-SN” folder. The “results” folder will contain three result files for RNA-seq sample: 
(1) the “SN-alter-3UTR.fasta” file is a fasta format file containing all the alternative 3’UTR sequences identified; 
(2) the “SN.change.txt” file contains the following information in each column: 
1) TCONS ID: assembled transcript ID from Cufflinks;
2) transcripts ID: corresponding transcript ID in the annotation;
3) gene ID: corresponding gene ID in the annotation;
4) class code: the type of match between the Cufflinks transcripts and the reference transcript. A full description of class code meaning can be obtained from the link: http://cole-trapnell-lab.github.io/cufflinks/ cuffcompare/
5) alternative 3UTR length; 
6) annotated 3UTR length;  
7) transdecoder; 
8)+/- transdecoder; 
9) 3utr length transdecoder: columns 7, 8, and 9 represent the source, strand and length of predicted 3’UTR using transdecoder;
10) FPKM-replicate-1; 
11) FPKM-replicate-2; 
12) FPKM-replicate-3: columns 10, 11, 12 represent the expression levels of the alternative 3’UTR for RNA-seq replicate1, 2, and 3, respectively.
(3) the “SN-R1/2/3-3utr-counts-30.txt” file contains the following information for each column: 
1) chromosome: chromosomal location of the alternative 3’UTR;
2) alternative 3utr_start; 
3) alternative 3utr_end: columns 2 and 3 represent the start and end positions of the alternative 3’UTR. 
4) transcript ID: the corresponding transcript ID in the annotation;
5) +/-;  
6) number of reads mapped to the 3'UTR region, those transcripts with less than 30 reads have been filtered out; 

2.	The “2nd-SN” folder contains all the secondary structure plots that are predicted for the priUTR- identified alternative 3’UTR sequences by RNAfold from the ViennaRNA package with minimum free energy.
